base/sparse/higherorderfns.jl

```diff
@@ -1061,8 +1062,8 @@ promote_containertype(::Type{Array}, ::Type{AbstractSparseArray}) = PromoteToSpa
 # mostly just disambiguates Array from the main containertype promotion mechanism
 # AbstractArray serves as a marker to shunt to the generic AbstractArray broadcast code
 _spcontainertype(x) = _containertype(x)
-_spcontainertype(::Type{<:Vector}) = Vector
-_spcontainertype(::Type{<:Matrix}) = Matrix
+_spcontainertype(::Type{<:AbstractVector}) = Vector
```

Comment:
In the future I'd love if we could have some kind of out-of-the-box functionality being guaranteed for types that implement some interfaces, e.g. the `AbstractArray` interface. That is, it would be cool IMO if you could properly define the interface methods and the you automatically you get the same behavior as for other subtypes of `AbstractArray` when broadcasting with other type of containers, and only have to worry about implementing more methods if you want a specific behavior.

This change and the general idea in #20102 are some proposals in that direction, but of course, that needs more careful design. So I could go with a more conservative approach for now, but I really hope we can come up with a better structure for all of these.